feat(nav): added horizontal subnav

mcoker opened the pull request:
fixes https://github.com/patternfly/patternfly/issues/3935

patternfly-build commented:
Preview: https://patternfly-pr-4229.surge.sh

mattnolting reviewed (commented):
Great work! 

**Observations**
- Looks the demos screenshots need to be generated. 
- The design specifies top border by default, but I think your approach is good. Maybe a bordered variant would be more versatile.

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -134,10 +137,40 @@
+  --pf-c-nav--m-horizontal-subnav__link--PaddingRight: var(--pf-global--spacer--md);
+  --pf-c-nav--m-horizontal-subnav__link--PaddingBottom: var(--pf-global--spacer--sm);
+  --pf-c-nav--m-horizontal-subnav__link--PaddingLeft: var(--pf-global--spacer--md);
+  --pf-c-nav--m-horizontal-subnav__link--xl--PaddingTop: var(--pf-global--spacer--md);
```
I was just matching the breakpoint where the vertical nav hides, horizontal/masthead nav links get smaller, so they change at the same time. @mcarrano @mceledonia wdyt?

You can see the demos here:
* https://patternfly-pr-4229.surge.sh/components/navigation/html-demos/horizontal-subnav/
* https://patternfly-pr-4229.surge.sh/components/navigation/html-demos/horizontal-nav-with-horizontal-subnav/

<img width="1321" alt="Screen Shot 2021-07-26 at 10 21 08 AM" src="https://user-images.githubusercontent.com/35148959/127015007-a2447373-e076-420d-8ee9-3d8df116578c.png">


<img width="1305" alt="Screen Shot 2021-07-26 at 10 21 12 AM" src="https://user-images.githubusercontent.com/35148959/127015027-67723e9b-a28a-4180-8bae-2a48d5c5a0e0.png">

mcoker commented on src/patternfly/components/Nav/nav.scss:
```diff
@@ -66,6 +66,7 @@
   --pf-c-nav__link--m-current--before--BorderBottomWidth: 0;
 
   // Link ::after borders
+  --pf-c-nav__link--after--Content: "";
```
Nope, just easier for us to use a var instead of writing out the selector. Reverted it though since it isn't intended for a user to use.

mcoker commented on src/patternfly/demos/Page/page-template-horizontal-subnav-nav.hbs:
```diff
@@ -0,0 +1,31 @@
+    {{#> nav-list}}
+      {{#> nav-item}}
+        {{#> nav-link nav-link--href="#" nav-link--current="true"}}
+          Tertiary nav item 1
```
Oops!

mcoker commented:
@mattnolting updated

> Looks the demos screenshots need to be generated.

I'll update all of them after code freeze when the release notes are generated.

> The design specifies top border by default, but I think your approach is good. Maybe a bordered variant would be more versatile.

Discussed offline, we can add this if it comes up for cases where a user wants to add it somewhere else in the page with these custom border colors.

mattnolting reviewed (approved):
LPTM!

mcarrano commented:
This is looking great @mcoker . Can you also review @mceledonia ?

My only question is why the nav items look unnecessarily tall what I look at them in the examples page.

![Screen Shot 2021-07-27 at 9 35 57 AM](https://user-images.githubusercontent.com/6619860/127163560-759cda0e-4f5e-4508-8ee7-c2dcf68cba3f.png)

Is this just an artifact of the workspace? The demos look correct.

mcoker commented:
> Is this just an artifact of the workspace?

@mcarrano yep, the example boxes have extra padding so that the things in the box don't touch the edges. Here is an outline of the actual component in the example box:

<img width="984" alt="Screen Shot 2021-07-27 at 9 17 58 AM" src="https://user-images.githubusercontent.com/35148959/127170213-96459563-6857-42ce-aaf2-4fceaf18f588.png">
<img width="886" alt="Screen Shot 2021-07-27 at 9 17 49 AM" src="https://user-images.githubusercontent.com/35148959/127170217-3e08e3a2-0c13-4caf-8491-9a55304928b9.png">

mcarrano commented:
One additional comment about the demos @mcoker . I think that we should actually remove the existing Tertiary nav demo and replace it with this one: https://patternfly-pr-4229.surge.sh/components/navigation/html-demos#horizontal-subnav calling it Tertiary nav as this is the pattern I think we want to promote going forward. The idea would be to only maintain support for the existing Tertiary nav for legacy purposes, but once this new nav is available, recommend that as the preferred solution whenever you want to have a sub-nav at the top of the page,

Does that make sense? Any problems with that approach? There is a related issue I opened here https://github.com/patternfly/patternfly/issues/3936 But I think this PR may cover it.

mcoker commented:
@mcarrano hmm... would we also remove the component example (not just the demo), and rename the horizontal subnav to tertiary nav there, too? If so, I think that works. I imagine we might rename the existing tertiary nav example/demos to "Legacy tertiary nav", so that we still have examples to test since tertiary nav is still technically supported.

mcarrano commented:
I agree that we should rename the existing Tertiary nav example to something like "legacy" rather than remove it entirely. I'm not really comfortable with calling the new variant "tertiary" since it could also be used for secondary navigation. In general, I never liked the term tertiary because it seems too prescriptive. I think we can also provide more recommendations in the design docs about when to use each variation.

mcoker commented:
@mcarrano updated, also updated the screenshots for all the demos since those would still show the old tertiary nav.

Demos I updated that now have the horizontal subnav instead of the tertiary nav.
* Nav
* Alert
* Page
* Data list
* Table

mcarrano commented:
This is great @mcoker . Thanks for updating all of those demos! The only problem I see is that on the Navigation demos page: https://patternfly-pr-4229.surge.sh/components/navigation/html-demos  The new demos are showing handlebar code where I should be seeing the screenshot.

![Screen Shot 2021-07-28 at 3 15 18 PM](https://user-images.githubusercontent.com/6619860/127382756-e6fba011-5009-4834-8bef-18d5c64556a1.png)

mcoker commented:
@mcarrano updated

mcarrano commented:
@mcoker the problem above is fixed, but still missing the screenshots. Otherwise looks great.

mcoker commented:
Thanks @mcarrano not sure why the screenshots are still broken, I'll verify they're correct when we build the release notes.

patternfly-build commented:
:tada: This PR is included in version 4.125.0 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v4.125.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/4.125.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
